### PR TITLE
Feature TAO-8910 keyboard navigation

### PR DESCRIPTION
**Pull request by bziondik**
Related to: https://oat-sa.atlassian.net/browse/TAO-8910

Was done:
added to jumplinks "Jump to: Navigation" and "Jump to: Question"
added focus styling to all buttons, links
removed all tabindex's an inside item
changed markup for proper navigation `views/js/review/plugins/navigation/next-prev/buttons.js `
added unit tests `views/js/test/review/provider/qtiTestReviewProvider/test.js `

### CODE REVIEW

File: views/js/review/plugins/content/item-answer/scss/item-answer.scss [hunk 1]
```diff
@@ -161,9 +163,23 @@ $incorrectColor: blacken($error, .2);
...
+        li.active .navigable {
+            &:focus {
+                outline-offset: -$focusActiveSpace;
+            }
+        }
+        
     }
 }
+// for navigation
+.qti-itemBody {
```
Comment (jsconan) [inline views/js/review/plugins/content/item-answer/scss/item-answer.scss]: Since this is not related to the plugin, it should be put on the main layout.

File: views/js/review/plugins/navigation/review-panel/scss/panel.scss [hunk 2]
```diff
@@ -227,7 +228,6 @@ $incorrectColor: $error;
 
     .review-panel-content {
         flex: 0 1 auto;
-        overflow-y: auto;
```
Comment (jsconan) [inline views/js/review/plugins/navigation/review-panel/scss/panel.scss]: Why is this removed? The component is now broken when the amount of elements make the panel bigger than the available space.

File: views/js/review/plugins/content/item-answer/plugin.js [hunk 3]
```diff
@@ -140,6 +140,10 @@ define([
                         } else {
                             itemAnswer.setIncorrect();
                         }
+                        // remove all tabindex's inside item for right navigation
+                        testRunner.getAreaBroker().getContentArea().find("[tabindex]").attr('tabindex', -1);
```
Comment (jsconan) [inline views/js/review/plugins/content/item-answer/plugin.js]: please use simple quotes

File: views/js/review/plugins/content/item-answer/plugin.js [hunk 4]
```diff
@@ -140,6 +140,10 @@ define([
...
+                        // add tabindex on itemBody
+                        testRunner.getAreaBroker().getContentArea().find(".qti-itemBody").attr('tabindex', 0);
```
Comment (jsconan) [inline views/js/review/plugins/content/item-answer/plugin.js]: Since the same sequence is used twice please consider to use a variable instead:
```javascript
const $contentArea = testRunner.getAreaBroker().getContentArea();
$contentArea.find('[tabindex]').attr('tabindex', -1);
$contentArea.find('.qti-itemBody').attr('tabindex', 0);
```

File: views/js/review/provider/qtiTestReviewProvider.js [hunk 5]
```diff
@@ -139,6 +140,28 @@ define([
 
             this.assetManager = assetManagerFactory();
 
+            // set tabindex=0 in navigation links and hide other links from navigation
+            $(document).find("li a:not([tabindex])").attr("tabindex", 0);
+            $(document).find("a:not([tabindex])").attr("tabindex", -1);
```
Comment (jsconan) [inline views/js/review/provider/qtiTestReviewProvider.js]: Please avoid to target the whole page, prefer scope instead to avoid side effects. Remember the component might be integrated in any place. So it should not interfere with the surrounding.

File: views/js/review/provider/scss/qtiTestReviewProvider.scss [hunk 6]
```diff
@@ -5,3 +5,7 @@
 @import "inc/section-container";
 @import "inc/test-dimensions";
 @import 'inc/review-layout';
+
+header nav ul li a[tabindex]:focus, .focused {
```
Comment (jsconan) [inline views/js/review/provider/scss/qtiTestReviewProvider.scss]: Please avoid to rely on tags. Always prefer classes.
Moreover, please scope the rules to the content belonging to the component, not the entire page.

File: views/js/review/provider/tpl/qtiTestReviewProvider.tpl [hunk 7]
```diff
@@ -1,4 +1,6 @@
 <main class="test-review-scope">
+    <a class="jumplink navigationlink" href="#" tabindex="1">{{__ "Jump to:"}} <b>{{__ "Navigation"}}</b></a>
+    <a class="jumplink answerlink" href="#" tabindex="1">{{__ "Jump to:"}} <b>{{__ "Question"}}</b></a>
```
Comment (jsconan) [inline views/js/review/provider/tpl/qtiTestReviewProvider.tpl]: I don't get the point for those jumplinks. Why not use anchors instead?

File: views/js/review/provider/qtiTestReviewProvider.js [hunk 8]
```diff
@@ -139,6 +140,28 @@ define([
...
+                });
+                link.on('focus', () => {
+                    areaBroker.getArea(where).addClass('focused');
+                });
+                link.on('blur', () => {
+                    areaBroker.getArea(where).removeClass('focused');
+                });
+            }
+            createJumplink(areaBroker.getContainer().find('.navigationlink'), 'sidebar');
+            createJumplink(areaBroker.getContainer().find('.answerlink'), 'contentWrapper');
```
Comment (jsconan) [inline views/js/review/provider/qtiTestReviewProvider.js]: Why didn't you use the `ui/keyNavigation` instead?
Moreover, please avoid to set per element event listeners. Always prefer delegated events.

File: views/js/review/plugins/content/item-answer/scss/item-answer.scss (hunk 1 above)
Comment (bziondik) [reply views/js/review/plugins/content/item-answer/scss/item-answer.scss]: After discussion with UX decided to remove focusing on the item.

File: views/js/review/plugins/content/item-answer/plugin.js (hunk 3 above)
Comment (bziondik) [reply views/js/review/plugins/content/item-answer/plugin.js]: fixed

File: views/js/review/provider/qtiTestReviewProvider.js (hunk 5 above)
Comment (bziondik) [reply views/js/review/provider/qtiTestReviewProvider.js]: removed

File: views/js/review/provider/tpl/qtiTestReviewProvider.tpl (hunk 7 above)
Comment (bziondik) [reply views/js/review/provider/tpl/qtiTestReviewProvider.tpl]: Because of the need to apply a style to an area when jumplink is in focus
And after click the need to set focus on the first navigatable element

File: views/js/review/provider/qtiTestReviewProvider.js (hunk 8 above)
Comment (bziondik) [reply views/js/review/provider/qtiTestReviewProvider.js]: Refactored. Now uses delegated events.

**Comment by bziondik** [comment]
@jsconan  I tried to apply ui/keyNavigation and can't set focus on any button. Focus is set on checkboxes/inputs in read-only item. And also this plugin blocked build-in browser navigation to other tabs from this tab.
So I decided not to use ui/keyNavigation, because of differences in common testRunner and this case.
We don't have buttons in bottom-action-bar, but in navigation panel, they exist. So I will need to change initNavigatorNavigation function
https://github.com/oat-sa/tao-test-runner-qti-fe/blob/428e1d8b597454ddec4b34980624146a6ce2f262/src/plugins/content/accessibility/keyNavigation.js#L126
Then main part about navigation in content part is not used because in this case it is readonly
Then initHeaderNavigation has a selector that doesn't match with the logout button
https://github.com/oat-sa/tao-test-runner-qti-fe/blob/428e1d8b597454ddec4b34980624146a6ce2f262/src/plugins/content/accessibility/keyNavigation.js#L102

File: views/js/review/plugins/content/item-answer/plugin.js (hunk 4 above)
Comment (bziondik) [reply views/js/review/plugins/content/item-answer/plugin.js]: fixed

File: views/js/review/provider/scss/qtiTestReviewProvider.scss (hunk 6 above)
Comment (bziondik) [reply views/js/review/provider/scss/qtiTestReviewProvider.scss]: removed

File: views/js/review/plugins/navigation/review-panel/scss/panel.scss (hunk 2 above)
Comment (bziondik) [reply views/js/review/plugins/navigation/review-panel/scss/panel.scss]: restored

File: views/js/review/provider/qtiTestReviewProvider.js [hunk 9]
```diff
@@ -139,6 +140,31 @@ define([
 
             this.assetManager = assetManagerFactory();
 
+            // first and second tab will show block and navigate to panel or content
+            const createJumplinks = (container) => {
+                container.on('click', (e) => {
+                    if ($(e.target).hasClass('jumplink')){
+                        e.preventDefault();
+                        $(e.target).blur();
+                        const where = $(e.target).data('area');
```
Comment (jsconan) [inline views/js/review/provider/qtiTestReviewProvider.js]: No need to use jQuery here:
```javascript
if (e.target.classList.contains('jumplink')){
    e.preventDefault();
    e.target.blur();
    const where = e.target.dataset.area;
```

At least you should store the jQuery object in a variable to avoid reconstruction on each line

File: views/js/review/provider/qtiTestReviewProvider.js [hunk 10]
```diff
@@ -139,6 +140,31 @@ define([
...
+                        areaBroker.getArea(where).find(":not(.hidden)[tabindex]").first().focus();
+                    }
+                });
+                container.on('focusin', (e) => {
+                    if ($(e.target).hasClass('jumplink')){
+                        const where = $(e.target).data('area');
```
Comment (jsconan) [inline views/js/review/provider/qtiTestReviewProvider.js]: same here, `jQuery` is not needed.

File: views/js/review/provider/qtiTestReviewProvider.js [hunk 11]
```diff
@@ -139,6 +140,31 @@ define([
...
+                        areaBroker.getArea(where).addClass('focused');
+                    }
+                    
+                });
+                container.on('focusout', (e) => {
+                    if ($(e.target).hasClass('jumplink')){
+                        const where = $(e.target).data('area');
```
Comment (jsconan) [inline views/js/review/provider/qtiTestReviewProvider.js]: here too, `jQuery` is not needed.

File: views/js/review/provider/qtiTestReviewProvider.js (hunk 9 above)
Comment (bziondik) [reply views/js/review/provider/qtiTestReviewProvider.js]: removed jQuery

File: views/js/review/provider/qtiTestReviewProvider.js (hunk 11 above)
Comment (bziondik) [reply views/js/review/provider/qtiTestReviewProvider.js]: removed jQuery

File: views/js/review/provider/qtiTestReviewProvider.js (hunk 10 above)
Comment (bziondik) [reply views/js/review/provider/qtiTestReviewProvider.js]: removed jQuery
